### DISCUSSION TITLE
Read out of already defined keywords

### DISCUSSION THREAD

**Discussion by JanKirner** [ansys/pydyna#478] [Q&A]
Hello everybody, 

As part of the preparation process of solver input files (.k-file) for LS-Dyna, I would like to use PyDyna to read out and modify existing solver input files. Some conspicuousness and questions have arisen. The import as a _DynaSolution_ object works without any problems using the _open_files_ method. 

In a minimal example for a solver input file, *PART_SOLID keywords have been defined among others, which means that also the keyword *NODES exists. The model attribute oft the _DynaSolution_ object  is updated using the __sync_up_model_ method. The __parts_ attribute of the model object is described with values, but the __solids_ and __nodes_ attributes are not. What is the reason for this behaviour?

As far as I can see from the minimal example, keywords already defined in the solver input file such as *DEFINE_COORDINATE_SYSTEM, *DEFINE_CURVE or *SET_NODE_LIST cannot be read and processed using PyDyna. However, it is possible to define those keywords using suitable classes or methods such as the method _create_general_keyword_. Is there a way to read and process existing keywords from a solver input file? From my point of view, this is an essential function.  

Thank you in advance.

**Comment by RobPasMue** [comment]
Pinging @ansys/pydyna for visibility

**Reply by koubaa** [reply]
@JanKirner this is now possible using the `keywords` module which was recently added. Please take a look at our updated documentation.

**Reply by ebzschulth** [reply]
Can anyone tell my where I can find this operation? I have been looking at most of the examples and https://github.com/ansys/pydyna/blob/main/tests/test_keywords.py, but could not find it unfortunately.

Thank You very much!

**Reply by koubaa** [reply]
@ebzschulth 

Something like this?

```
def modify_deck(deck):
    """ modify the keywords in the deck."""
    # REPLACE THE BELOW WITH THE MODIFICATION YOU WANT TO DO!
    section_shells = deck.get_kwds_by_full_type("SECTION", "SHELL")
    for section_shell in section_shells:
         section_shell.secid = -1

deck =Deck()
deck.import_file('/path/to/old.k')

modify_deck(deck)
deck.export_file('/path/to/new.k')
```

**Reply by ebzschulth** [reply]
@koubaa, thanks for your response. 

As I was studying the docs I noticed, that I am looking for a way to change the values below a keyword, called cards.

Lets say, I have a keyword 
`Keyword: *MAT_ELASTIC
$#     mid        ro         e        pr        da        db    unused
         1  0.00025910000000.0      0.33`

I was trying to change the parameter "mid" using 
`for keyword in deck.keywords:
    print(f"Keyword: {keyword}")
    for card in keyword.cards:
        if keyword == "*MAT_ELASTIC":
           card.__setattr__(name="mid", value=2)`

but I can't change it. Where am I wrong here? What is the correct way to set new values for a card?

Thank you very much.

**Reply by koubaa** [reply]
@ebzschulth you can just use `keyword.mid=2`

**Reply by ebzschulth** [reply]
Thanks for your quick response, it works!

**Reply by ebzschulth** [reply]
Hi @koubaa, is there also a way to change the parameters in a *.dyn file? I was trying to read it the same way:

```
# Open deck
deck = Deck()
deck.import_file(sim_dyn_file_path)
    
# Change keyword value
for keyword in deck.keywords:
           keyword.__setattr__(variable_name, float(new_value)
           
           
which doesnt work unfortunately.
```

Thanks

**Reply by koubaa** [reply]
@ebzschulth pydyna doesn't look at the file extension, as long as the .dyn file follows the same keyword file spec, it should load it in the same way. Please share a minimal example reproducing the problem if you would like us to look into it

**Reply by ebzschulth** [reply]
I get the exception: Exception: Only comments can precede *KEYWORD

Here is a minimal example, where I am trying to change the parameter U to 0.24:
[change_process_params.zip](https://github.com/user-attachments/files/20369688/change_process_params.zip)

The file path in change_process_params_minimal_example has to be changed to the file sim.dyn.
